sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala

```diff
@@ -211,37 +264,69 @@ trait StreamTest extends QueryTest with Timeouts {
               })
 
           case StopStream =>
-            checkState(currentStream != null, "can not stop a stream that is not running")
-            currentStream.stop()
-            currentStream = null
+            assert(currentStream != null, "can not stop a stream that is not running")
+            try failAfter(streamingTimout) {
+              currentStream.stop()
+              assert(!currentStream.microBatchThread.isAlive,
+                s"microbatch thread not stopped")
+              assert(!currentStream.isActive,
+                "query.isActive() is false even after stopping")
+              assert(currentStream.exception.isEmpty,
+                "query.exception() is not empty after clean stop")
+            } catch {
+              case _: InterruptedException =>
+              case _: org.scalatest.exceptions.TestFailedDueToTimeoutException =>
+                failTest("Timed out while stopping and waiting for microbatchthread to terminate.")
+              case t: Throwable =>
+                failTest("Error while checking stream failure.", t)
+            } finally {
+              lastStream = currentStream
+              currentStream = null
+            }
 
           case DropBatches(num) =>
-            checkState(currentStream == null, "dropping batches while running leads to corruption")
+            assert(currentStream == null, "dropping batches while running leads to corruption")
             sink.dropBatches(num)
 
-          case ExpectFailure =>
+          case ef: ExpectFailure[_] =>
+            assert(currentStream != null, "dropping batches while running leads to corruption")
             try failAfter(streamingTimout) {
-              while (streamDeathCause == null) {
-                Thread.sleep(100)
+              val thrownException = intercept[QueryException] {
+                currentStream.awaitTermination()
+              }
+              eventually("microbatch thread not stopped after termination with failure") {
```

Comment:
do we want to provide the guarantee that the thread has stopped by the time awaitTermination unblocks?